### DISCUSSION TITLE
`bitmap` texture to visualize public transportation data?

### DISCUSSION THREAD

**Discussion by charnould** [visgl/deck.gl#6986] [Q&A]
Hi,

First, thanks for making `deck.gl`, it's truly an impressive library (like `H3` and `kepler` by the way)!
I'm pretty new to `deck.gl`, a self taught pretty bad dev and I'd like to use `deck.gl` to show public transportation/transit for region of Paris (France) using `GTFS` data (a well known open source specification).

At this time:

1) I've made a `python` script to transform `GTFS` data for region of Paris in a [format `deck.gl` understand](https://raw.githubusercontent.com/charnould/reclus/master/gtfs.json) (caution: 25 mb `json` file) and more precisely `trip-layer`. Problem: this file is already quite heavy (25 mb) and only contains a tiny part of my data. The real file is 250 mb large.

2) Map is however functional (not using `react` by choice): https://charnould.github.io/reclus/

Because (i) I can't ask visitors to download a 250 mb file to see the full map and (ii) animation isn't smooth at all (!) with this huge file (it's unusable), I'm thinking to an alternative idea, inspired by this [amazing wind visualization](https://philogb.github.io/page/wind/).

The idea: 

- Create a `texture` `bitmap` where each pixel is a vehicle (a train, subway, bus...).  
`RGBA` of this pixel is the lat/long of this vehicle at a specific time
- Draw 288 bitmaps representing a 10 minute-interval (24 hours x 60 minutes / 10 minutes = 288)
- use `deck.gl` and these `bitmaps` to render a smooth animation...

My questions:
- Is this the right road (pun intended)?
- Do you have any clue to get started?  
(I don't know where to look to generate these `bitmaps` and use `deck.gl` to render it.

Thanks a lot.

**Comment by kylebarron** [comment]
For another data point, I did a similar project a few years ago: https://github.com/kylebarron/all-transit. The idea was to use deck.gl's `TripLayer` _within a `TileLayer`_, so that each individual tile would render a moving "trail". It worked really well! ... And then I was cleaning up my S3 bucket and accidentally deleted the prod data 🤦‍♂️ . So there isn't a live demo you can still access to see the moving vehicles, but this image shows a still frame. All the orange flashes are moving trails using the `TileLayer`.

<img src="https://user-images.githubusercontent.com/15164633/172233848-99ad1736-5e05-4568-bb0e-165fdbad5f6c.png" height="400" />

In general, using a `TripLayer` vs a collection of bitmaps is probably a question of how many pixels would have data points. Paths of moving vehicles are more likely to be sparse, so a vector encoding then using the `TripLayer` would likely be more space efficient than a data cube of bitmaps.

**Reply by charnould** [reply]
Kyle,  
First, thank for your reply.  
However, even if I read thoroughly your comment and [blog post](https://kylebarron.dev/blog/all-transit), I'm not sure I understand it fully.  
Could you please say me if my understanding is good:
- To make my map smooth, I need to split my huge data file into small ones called [`tiles`](https://github.com/visgl/deck.gl/blob/8.7-release/docs/api-reference/geo-layers/tile-layer.md)
- Depending on my current viewport (when navigating in my map in the browser), `tileLayer` will automatically callback `getTileData` to fetch the small corresponding `json` dataset
- Then, I use `tripLayer` to animate data contained in this small dataset.

But I don't understand how to generate this tiny json files (`tiles`) from my huge one. 
Do I need to use [`tippecanoe`](https://github.com/mapbox/tippecanoe) or another tool?
Does the new service proposed by Mapbox ([`Mapbox Tiling Service`](https://docs.mapbox.com/mapbox-tiling-service/guides/)) of any use considering my objective?
Thanks again.

PS: With `All transit` you seem to use `GTFS` data. However How can you make - for example - a bus/tramway follows the road geometry when `GTFS` contains only X/Y stops...?  
In my map, the animation is a straight line between two stops and [doesn't follow at all the road](https://charnould.github.io/reclus/)...

**Reply by kylebarron** [reply]
First, not saying you _shouldn't_ do your original approach; just saying this is how I tackled the problem.

> * To make my map smooth, I need to split my huge data file into small ones called [`tiles`](https://github.com/visgl/deck.gl/blob/8.7-release/docs/api-reference/geo-layers/tile-layer.md)
> * Depending on my current viewport (when navigating in my map in the browser), `tileLayer` will automatically callback `getTileData` to fetch the small corresponding `json` dataset
> * Then, I use `tripLayer` to animate data contained in this small dataset.

Yeah that's right.

> But I don't understand how to generate this tiny json files (`tiles`) from my huge one. 

That was the majority of the work in https://github.com/kylebarron/all-transit. The [README](https://github.com/kylebarron/all-transit#data) is very, very descriptive for how I created the data.

> PS: With `All transit` you seem to use `GTFS` data. However How can you make - for example - a bus/tramway follows the road geometry when `GTFS` contains only X/Y stops...?

I actually used transitland data (https://www.transit.land/). Which is essentially upgraded, reprocessed GTFS data. But also GTFS has a [shapes file](https://gtfs.org/schedule/reference/#shapestxt). You can join this file to other files to get the shape of each route.
